### ISSUE TITLE
CPU Usage High When Backend Down

### ISSUE THREAD

**Issue by phihos** [haproxy/haproxy#1902]
### Detailed Description of the Problem

This issue is somewhat related to #1770 which has been opened by me a while ago, but the config is different and maybe the underlying issue too. This is why I create a new issue.

Since last time the request volume in my haproxy instance has risen from ~10k reqs/s to ~20reqs/s. When the load climbs over ~17k reqs/s L6 health checks begin to time out although the backend server is fine. But this is not the actual issue here since this may also be a OS/networking problem. The problem that is definitely linked to haproxy is that as soon as one backend is declared down the CPU usage climbs from 40-50% to 100% and that seems to cause all other backends to be declared down as well. 
The run queue contains a lot of items at that time (rises from ~30 to ~140k) so I guess the health checks for other servers are stuck in the queue and do not get executed.

Here we can see a timeline of the checks and we can see that all but one backend have failed health checks.
![grafik](https://user-images.githubusercontent.com/1387322/196509619-4d9d9448-dc7f-4d32-99b1-e97dff97da08.png)
All downed backends are have a single HTTPS server with a L6 active health check. The only backend that is up has a non-SSL HTTP server with a L4 check.

I remembered your advice from last time to use `perf top` when it happens and took a few:

This is when everything is fine as baseline:
![grafik](https://user-images.githubusercontent.com/1387322/196506944-4bb1cd5b-df6a-4a0d-90a5-4c8c5eacc052.png)

And this is after all backends are down (I made a few on multiple occasions when this happened): 
![grafik](https://user-images.githubusercontent.com/1387322/196507152-776e0ed3-17c9-4dac-b6ed-6d2f4a841116.png)
![grafik](https://user-images.githubusercontent.com/1387322/196507288-cf38514d-7801-4479-91e9-59c86ea27494.png)
![grafik](https://user-images.githubusercontent.com/1387322/196507403-c872f952-4f3c-4f89-aa9c-e979d097c467.png)
![grafik](https://user-images.githubusercontent.com/1387322/196507504-a8363404-2162-429b-ba33-cc38816066e5.png)

Also one time haproxy crashed:
```
Thread 24 is about to kill the process.
  Thread 1 : id=0x7f1da7235f80 act=1 glob=0 wq=1 rq=1 tl=0 tlsz=-20086 rqsz=3
      1/1    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=11778469228373 now=11778469520674 diff=292301
             curr_task=0x555a638a1d20 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x555a12d97aa0
  Thread 2 : id=0x7f1da722a700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-19942 rqsz=0
      1/2    stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=11210976659022 now=11210976710464 diff=51442
             curr_task=0
  Thread 3 : id=0x7f1da68a4700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-20029 rqsz=0
      1/3    stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=10551333155286 now=10551333253586 diff=98300
             curr_task=0
  Thread 4 : id=0x7f1da60a3700 act=1 glob=1 wq=1 rq=0 tl=0 tlsz=-20285 rqsz=0
      1/4    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10380779644726 now=10380784566978 diff=4922252
             curr_task=0x7f195cc4ffa0 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1b913ae640
  Thread 5 : id=0x7f1da58a2700 act=1 glob=0 wq=1 rq=0 tl=0 tlsz=-20295 rqsz=0
      1/5    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10016464036550 now=10016464412270 diff=375720
             curr_task=0x7f1c613bb030 (tasklet) calls=36
               fct=0x5559d71abc70(h2_io_cb) ctx=0x7f1a6a115510
  Thread 6 : id=0x7f1da50a1700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-20238 rqsz=0
      1/6    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9908277739757 now=9908277973737 diff=233980
             curr_task=0
  Thread 7 : id=0x7f1da48a0700 act=1 glob=0 wq=1 rq=1 tl=1 tlsz=-20264 rqsz=15
      1/7    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10895342611689 now=10895343071800 diff=460111
             curr_task=0x7f1adeca3810 (tasklet) calls=1
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f17d1f1bde0
  Thread 8 : id=0x7f1d20348700 act=1 glob=0 wq=1 rq=0 tl=0 tlsz=-20356 rqsz=0
      1/8    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9770498384976 now=9770498686376 diff=301400
             curr_task=0x7f1a430e3c20 (tasklet) calls=289
               fct=0x5559d71abc70(h2_io_cb) ctx=0x7f18ab44c490
  Thread 9 : id=0x7f1d1fb47700 act=1 glob=0 wq=1 rq=0 tl=0 tlsz=-19841 rqsz=0
      1/9    stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9955034355906 now=9955034769915 diff=414009
             curr_task=0x7f1794538cf0 (tasklet) calls=3
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1b109e0c50
  Thread 10: id=0x7f1d1f346700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-20012 rqsz=0
      1/10   stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=9980051115066 now=9980051340133 diff=225067
             curr_task=0
  Thread 11: id=0x7f1d1eb45700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-20153 rqsz=0
      1/11   stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=9873140510663 now=9873141087832 diff=577169
             curr_task=0
  Thread 12: id=0x7f1d1e344700 act=1 glob=1 wq=1 rq=0 tl=0 tlsz=-20629 rqsz=0
      1/12   stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=9892513240206 now=9892513663474 diff=423268
             curr_task=0
  Thread 13: id=0x7f1d1db43700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-20418 rqsz=0
      1/13   stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=9951687650317 now=9951688109110 diff=458793
             curr_task=0
  Thread 14: id=0x7f1d1d342700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-19887 rqsz=0
      1/14   stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=9871971403513 now=9871971857175 diff=453662
             curr_task=0
  Thread 15: id=0x7f1d1cb41700 act=1 glob=1 wq=1 rq=0 tl=1 tlsz=-19587 rqsz=3
      1/15   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9879825739594 now=9879829916837 diff=4177243
             curr_task=0x7f1a500e3f50 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f18983e6950
  Thread 16: id=0x7f1d1c340700 act=1 glob=1 wq=1 rq=0 tl=1 tlsz=-19512 rqsz=16
      1/16   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9773682473861 now=9773684423882 diff=1950021
             curr_task=0x7f143344ae60 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1bedbc98d0
  Thread 17: id=0x7f1d1bb3f700 act=1 glob=1 wq=1 rq=0 tl=1 tlsz=-20523 rqsz=1
      1/17   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9778068889786 now=9778073742788 diff=4853002
             curr_task=0x7f1be332d450 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f17b7e885e0
  Thread 18: id=0x7f1d1b33e700 act=1 glob=1 wq=1 rq=0 tl=1 tlsz=-20375 rqsz=4
      1/18   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9846497959330 now=9846504120351 diff=6161021
             curr_task=0x7f1cae99e4f0 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1a96ee5170
  Thread 19: id=0x7f1d1ab3d700 act=1 glob=0 wq=1 rq=0 tl=1 tlsz=-20220 rqsz=3
      1/19   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9893962108490 now=9893962751360 diff=642870
             curr_task=0x7f1bbaa3d110 (tasklet) calls=3
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1c41303f20
  Thread 20: id=0x7f1d1a33c700 act=1 glob=0 wq=1 rq=1 tl=1 tlsz=-20599 rqsz=2
      1/20   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10034567169639 now=10034567941259 diff=771620
             curr_task=0x7f1c1e57f970 (tasklet) calls=645
               fct=0x5559d71abc70(h2_io_cb) ctx=0x7f1721c8cd90
  Thread 21: id=0x7f1d19b3b700 act=1 glob=0 wq=1 rq=0 tl=0 tlsz=-20038 rqsz=11
      1/21   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9888830876596 now=9888831631697 diff=755101
             curr_task=0x7f19ea82ab50 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f19eb40a310
  Thread 22: id=0x7f1d1933a700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-19845 rqsz=0
      1/22   stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=9955314135170 now=9955315073840 diff=938670
             curr_task=0
  Thread 23: id=0x7f1d18b39700 act=1 glob=0 wq=1 rq=0 tl=0 tlsz=-19986 rqsz=0
      1/23   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10127955688535 now=10127956533685 diff=845150
             curr_task=0x7f1c2df2f990 (task) calls=1 last=0
               fct=0x5559d71e6d70(process_stream) ctx=0x7f1553c0f0f0
             strm=0x7f1553c0f0f0,800 src=162.158.90.182 fe=origin be=origin dst=unknown
             txn=0x7f188894bd70,40000 txn.req=MSG_BODY,d txn.rsp=MSG_RPBEFORE,0
             rqf=40d08002 rqa=30 rpf=80000000 rpa=0
             scf=0x7f12953ff9e0,EST,20 scb=0x7f1caf795350,INI,21
             af=(nil),0 sab=(nil),0
             cof=0x7f1701059fc0,a0000300:H2(0x7f1a2e9b59f0)/SSL(0x7f139c6bb4c0)/tcpv4(23417)
             cob=(nil),0:NONE((nil))/NONE((nil))/NONE(-1)
*>Thread 24: id=0x7f1d18338700 act=1 glob=1 wq=1 rq=1 tl=1 tlsz=-20417 rqsz=7
      1/24   stuck=1 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10307587669367 now=10310547931421 diff=2960262054
             curr_task=0x5559d8a69520 (task) calls=33864 last=0
               fct=0x5559d7212ee0(srv_cleanup_toremove_conns) ctx=(nil)
             call trace(13):
             | 0x7f1da7c02420 [48 c7 c0 0f 00 00 00 0f]: libpthread:+0x14420
             | 0x5559d7212f5c [48 89 d8 48 87 02 48 83]: srv_cleanup_toremove_conns+0x7c/0xcf
  Thread 25: id=0x7f1d17b37700 act=1 glob=1 wq=1 rq=0 tl=0 tlsz=-19907 rqsz=0
      1/25   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9893876561834 now=9893881000789 diff=4438955
             curr_task=0x7f16ba852380 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1c387a1890
  Thread 26: id=0x7f1d17336700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-20166 rqsz=0
      1/26   stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=9865344592055 now=9865345808307 diff=1216252
             curr_task=0
  Thread 27: id=0x7f1d16b35700 act=1 glob=0 wq=1 rq=0 tl=0 tlsz=-20039 rqsz=2
      1/27   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9733396540885 now=9733397882696 diff=1341811
             curr_task=0x7f1c29a857e0 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1b889270c0
  Thread 28: id=0x7f1d16334700 act=1 glob=0 wq=1 rq=0 tl=1 tlsz=-19938 rqsz=1
      1/28   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10159562819331 now=10159563511988 diff=692657
             curr_task=0x5559f009ffe0 (task) calls=1 last=0
               fct=0x5559d71e6d70(process_stream) ctx=0x555a231dbdb0
             strm=0x555a231dbdb0,800 src=172.68.110.189 fe=origin be=origin dst=unknown
             txn=0x555a6ac24b70,40000 txn.req=MSG_BODY,d txn.rsp=MSG_RPBEFORE,0
             rqf=40d08002 rqa=30 rpf=80000000 rpa=0
             scf=0x555a8b108e40,EST,20 scb=0x555a5e5d2a20,INI,21
             af=(nil),0 sab=(nil),0
             cof=0x7f15264c9f10,a0000300:H1(0x555a31a5b520)/SSL(0x555a09a1f150)/tcpv4(229109)
             cob=(nil),0:NONE((nil))/NONE((nil))/NONE(-1)
  Thread 29: id=0x7f1d15b33700 act=1 glob=0 wq=1 rq=0 tl=1 tlsz=-19978 rqsz=2
      1/29   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9826827973345 now=9826829469175 diff=1495830
             curr_task=0x7f192c44c060 (task) calls=3 last=0
               fct=0x5559d71e6d70(process_stream) ctx=0x7f1a1fecf0d0
             strm=0x7f1a1fecf0d0,80a src=172.68.254.163 fe=origin be=bk_2 dst=server_2:80
             txn=0x7f171ee2f800,40800 txn.req=MSG_CLOSED,d txn.rsp=MSG_CLOSED,d
             rqf=48c4e060 rqa=0 rpf=c004a063 rpa=0
             scf=0x7f158544c980,CLO,0 scb=0x7f1a517ff340,CLO,19
             af=(nil),0 sab=(nil),0
             cof=0x7f1682f7bf80,a0000300:H2(0x7f19bd8060f0)/SSL(0x7f19aaf8a9c0)/tcpv4(363569)
             cob=0x7f16a01cbb10,c0300:H1(0x7f1977abf7f0)/RAW((nil))/tcpv4(1971643)
  Thread 30: id=0x7f1d15332700 act=1 glob=0 wq=1 rq=0 tl=0 tlsz=-20120 rqsz=11
      1/30   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9751142749788 now=9751144176728 diff=1426940
             curr_task=0x7f1d01c8b080 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1b9d29b930
  Thread 31: id=0x7f1d14b31700 act=1 glob=0 wq=1 rq=0 tl=1 tlsz=-20171 rqsz=7
      1/31   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9797374101730 now=9797375689171 diff=1587441
             curr_task=0x7f17b8e47680 (task) calls=3 last=0
               fct=0x5559d71e6d70(process_stream) ctx=(nil)
  Thread 32: id=0x7f1d14330700 act=1 glob=0 wq=1 rq=1 tl=0 tlsz=-20337 rqsz=1
      1/32   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9708910434788 now=9708911949278 diff=1514490
             curr_task=0x7f1452d39410 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1aec298dd0
  Thread 33: id=0x7f1d13b2f700 act=1 glob=1 wq=1 rq=1 tl=1 tlsz=-20119 rqsz=3
      1/33   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9780131373782 now=9780137516344 diff=6142562
             curr_task=0x7f145ef40a00 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1cf9624640
  Thread 34: id=0x7f1d1332e700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-19938 rqsz=0
      1/34   stuck=0 prof=0 harmless=1 wantrdv=0
             cpu_ns: poll=9822143281590 now=9822144804312 diff=1522722
             curr_task=0
  Thread 35: id=0x7f1d109cb700 act=1 glob=0 wq=1 rq=0 tl=1 tlsz=-19696 rqsz=11
      1/35   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9783140055912 now=9783140207542 diff=151630
             curr_task=0x7f1bb2924980 (tasklet) calls=3
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1876229560
  Thread 36: id=0x7f1d101ca700 act=1 glob=0 wq=1 rq=0 tl=1 tlsz=-19806 rqsz=5
      1/36   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9826271937276 now=9826276755518 diff=4818242
             curr_task=0x7f17082ec120 (task) calls=1 last=0
               fct=0x5559d71e6d70(process_stream) ctx=0x7f1baf310270
             strm=0x7f1baf310270,800 src=172.70.246.184 fe=origin be=origin dst=unknown
             txn=0x7f19f0957480,0 txn.req=MSG_BODY,d txn.rsp=MSG_RPBEFORE,0
             rqf=40d08002 rqa=30 rpf=80000000 rpa=0
             scf=0x7f194023c350,EST,20 scb=0x7f1c58e49fd0,INI,21
             af=(nil),0 sab=(nil),0
             cof=0x7f138d406ce0,a0000300:H1(0x7f1c2c02d370)/SSL(0x7f1a660e8260)/tcpv4(1945405)
             cob=(nil),0:NONE((nil))/NONE((nil))/NONE(-1)
  Thread 37: id=0x7f1d0f9c9700 act=1 glob=0 wq=1 rq=0 tl=1 tlsz=-20635 rqsz=6
      1/37   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9902130260803 now=9902134885825 diff=4625022
             curr_task=0x7f184ab4ae70 (task) calls=1 last=0
               fct=0x5559d71e6d70(process_stream) ctx=0x7f1c4eb7ef50
             strm=0x7f1c4eb7ef50,800 src=172.68.186.172 fe=origin be=origin dst=unknown
             txn=0x7f190e3f9670,40000 txn.req=MSG_BODY,d txn.rsp=MSG_RPBEFORE,0
             rqf=d08002 rqa=30 rpf=80000000 rpa=0
             scf=0x7f1acccfca00,EST,20 scb=0x7f14ef385750,INI,21
             af=(nil),0 sab=(nil),0
             cof=0x555a1abcd840,a0000300:H2(0x7f147da4fcb0)/SSL(0x7f1778e6bcb0)/tcpv4(345299)
             cob=(nil),0:NONE((nil))/NONE((nil))/NONE(-1)
  Thread 38: id=0x7f1d0f1c8700 act=1 glob=1 wq=1 rq=0 tl=1 tlsz=-19903 rqsz=1
      1/38   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=10304512564060 now=10304518945702 diff=6381642
             curr_task=0x7f178a60d970 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f1989eb4750
  Thread 39: id=0x7f1d0e9c7700 act=1 glob=0 wq=1 rq=0 tl=0 tlsz=-19923 rqsz=0
      1/39   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9769402826223 now=9769406972944 diff=4146721
             curr_task=0x7f172d13ca50 (task) calls=3 last=0
               fct=0x5559d71e6d70(process_stream) ctx=0x7f12e20db010
             strm=0x7f12e20db010,10080a src=172.68.51.188 fe=origin be=bk_3 dst=server_3:443
             txn=0x7f146814b5b0,40000 txn.req=MSG_CLOSED,d txn.rsp=MSG_CLOSED,c
             rqf=4884e061 rqa=0 rpf=c004a063 rpa=0
             scf=0x7f18dfe5a940,CLO,0 scb=0x7f1b6f03a0a0,CLO,19
             af=(nil),0 sab=(nil),0
             cof=0x7f1bbf16a600,a0000300:H2(0x7f16346ea910)/SSL(0x7f138fefcac0)/tcpv4(1601361)
             cob=0x555a2fa2c9a0,300:H2(0x7f1748d6c840)/SSL(0x7f1830a11350)/tcpv4(1255924)
  Thread 40: id=0x7f1d0e1c6700 act=1 glob=1 wq=1 rq=0 tl=1 tlsz=-19722 rqsz=1
      1/40   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9805177497160 now=9805187192864 diff=9695704
             curr_task=0x7f1ab6a761c0 (tasklet) calls=2
               fct=0x5559d71540a0(ssl_sock_io_cb) ctx=0x7f17a8e9af80
  Thread 41: id=0x7f1d0d9c5700 act=0 glob=0 wq=1 rq=0 tl=0 tlsz=-20657 rqsz=0
      1/41   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9825810193663 now=9825812308624 diff=2114961
             curr_task=0
  Thread 42: id=0x7f1d0d1c4700 act=1 glob=0 wq=1 rq=1 tl=1 tlsz=-19826 rqsz=12
      1/42   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=12653915990149 now=12653920254021 diff=4263872
             curr_task=0x7f16b8cd70a0 (tasklet) calls=32539
               fct=0x5559d86fe380(main+0x15b9c50) ctx=0x7f1683853920
  Thread 43: id=0x7f1d0c9c3700 act=1 glob=0 wq=1 rq=0 tl=1 tlsz=-20092 rqsz=19
      1/43   stuck=0 prof=0 harmless=0 wantrdv=0
             cpu_ns: poll=9842177927839 now=9842182288431 diff=4360592
             curr_task=0x555a1097ca30 (task) calls=1 last=0
               fct=0x5559d71e6d70(process_stream) ctx=0x7f13a523aa50
             strm=0x7f13a523aa50,80a src=188.114.111.13 fe=origin be=bk_1 dst=server_1:444
             txn=0x7f19ff8d3800,40000 txn.req=MSG_DONE,d txn.rsp=MSG_RPBEFORE,0
             rqf=4cd08002 rqa=8000 rpf=80000000 rpa=0
             scf=0x7f158cd2d5c0,EST,20 scb=0x7f18aeac8dd0,ASS,31
             af=(nil),0 sab=(nil),0
             cof=0x7f1c23125dd0,a0000300:H2(0x7f1837e12a00)/SSL(0x7f1732b13720)/tcpv4(1099501)
             cob=0x555a0f4b9140,9c00000:NONE(0x7f18aeac8dd0)/SSL(0x7f1bca855f10)/NONE(-1)
[ALERT]    (1821) : Current worker (186545) exited with code 134 (Aborted)
[ALERT]    (1821) : exit-on-failure: killing every processes with SIGTERM
[WARNING]  (1821) : All workers exited. Exiting... (134)
```

I you need any more data about the incidents I am happy to help.

The last time this happened in #1770 the solution was to use `sc_` instead of `src_` fetches and a [fix](https://github.com/haproxy/haproxy/commit/8bd146d8af78371f97b66e50cac718666eb93388) that throtteled stick-table peer sync.

But this time I already use the `sc_` fetches and I do not sync big stick tables anymore. Yet the issue seems to revolve around stick-tables again. 

### Expected Behavior

Even with a big request volume the CPU load should not skyrocket when a backend is down. Also other backends should not be declared to be down as well.

### Steps to Reproduce the Behavior

1. Have a machine with a `AMD EPYC 7443` CPU.
2. Have over 17k reqs/s.
3. Have a flaky connection to a single backend.
4. When the healthcheck failed all other backends fail too.
5. CPU usage skyrockets at the same time. 

### What is your configuration?

```haproxy
global
  chroot  /var/lib/haproxy
  daemon  
  group  haproxy
  log  stdout len 4096 format raw local0 notice
  log  localhost:5140 len 4096 format rfc5424 local0 info
  maxconn  3500000
  nbthread  43
  pidfile  /var/run/haproxy.pid
  spread-checks  5
  ssl-default-bind-options  no-sslv3 no-tlsv10 no-tlsv11
  ssl-dh-param-file  /etc/haproxy/dhparams.pem
  stats  socket /var/lib/haproxy/stats
  tune.bufsize  34816
  tune.ssl.cachesize  3000000
  user  haproxy

defaults
  default-server  ca-file /etc/ssl/certs/ca-certificates.crt inter 10s fall 3 rise 2 maxqueue 30000
  http-reuse  always
  log  global
  option  redispatch
  option  contstats
  option  forwardfor
  option  httplog
  option  log-health-checks
  option  splice-auto
  option  tcp-smart-accept
  option  tcp-smart-connect
  option  abortonclose
  retries  3
  stats  enable
  timeout  http-request 10s
  timeout  http-keep-alive 30m
  timeout  connect 5s
  timeout  client 60s
  timeout  server 5s
  timeout  check 10s
  timeout  queue 10s
  timeout  client-fin 1s
  timeout  server-fin 1s

frontend origin
  bind *:8080 accept-proxy
  bind *:8443 accept-proxy  ssl crt /etc/haproxy/ssl alpn h2,http/1.1 ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256 ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384 tfo
  mode http
  
  
  # host and path acls for use_backend omitted...
  
  acl uri_too_long  url_len gt 32768
  acl is_blockedagent hdr_sub(user-agent) -i -f /etc/haproxy/maps/blacklist_user_agent.txt

  acl greylisted src -f /etc/haproxy/maps/greylisted_subnets.txt
  acl greylisted table_cnt(abuse) ge 20
  acl greylisted table_cnt(abuse-heavy) ge 20
  acl whitelisted src -f /etc/haproxy/maps/whitelisted_subnets.txt
  acl whitelisted req.fhdr(Authorization) -f /etc/haproxy/maps/whitelisted_auth_headers.txt
  acl is_json hdr_sub(Content-Type) -i application/json
  acl low_prio_queue sc_http_req_rate(1) ge <value omitted>
  acl low_prio_queue sc_http_req_rate(0) ge <value omitted>
  acl low_prio_queue_if_greylisted sc_http_req_rate(1) ge <value omitted>
  acl low_prio_queue_if_greylisted sc_http_req_rate(0) ge <value omitted>
  acl abuser sc_http_req_rate(1) ge <value omitted>
  acl abuser sc_http_req_rate(0) ge <value omitted>
  acl abuser sc_http_err_rate(1) ge <value omitted>
  acl abuser sc_http_err_rate(0) ge <value omitted>
  acl abuser sc_bytes_out_rate(1) ge <value omitted>
  acl abuser sc_bytes_out_rate(0) ge <value omitted>
  acl heavy_abuser sc_http_req_rate(1) ge <value omitted>
  acl heavy_abuser sc_http_req_rate(0) ge <value omitted>
  acl heavy_abuser sc_http_err_rate(1) ge <value omitted>
  acl heavy_abuser sc_http_err_rate(0) ge <value omitted>
  acl heavy_abuser sc_bytes_out_rate(1) ge <value omitted>
  acl heavy_abuser sc_bytes_out_rate(0) ge <value omitted>
  acl abuser_if_greylisted sc_http_req_rate(1) ge <value omitted>
  acl abuser_if_greylisted sc_http_req_rate(0) ge <value omitted>
  acl abuser_if_greylisted sc_http_err_rate(1) ge <value omitted>
  acl abuser_if_greylisted sc_http_err_rate(0) ge <value omitted>
  acl abuser_if_greylisted sc_bytes_out_rate(1) ge <value omitted>
  acl abuser_if_greylisted sc_bytes_out_rate(0) ge <value omitted>
  acl heavy_abuser_if_greylisted sc_http_req_rate(1) ge <value omitted>
  acl heavy_abuser_if_greylisted sc_http_req_rate(0) ge <value omitted>
  acl heavy_abuser_if_greylisted sc_http_err_rate(1) ge <value omitted>
  acl heavy_abuser_if_greylisted sc_http_err_rate(0) ge <value omitted>
  acl heavy_abuser_if_greylisted sc_bytes_out_rate(1) ge <value omitted>
  acl heavy_abuser_if_greylisted sc_bytes_out_rate(0) ge <value omitted>

  acl inc_abuse_cnt src_inc_gpc0(abuse) gt 0
  acl inc_abuse_heavy_cnt src_inc_gpc0(abuse-heavy) gt 0
  acl inc_abuse_but_whitelisted_cnt src_inc_gpc0(abuse-but-whitelisted) gt 0
  acl inc_abuse_heavy_but_whitelisted_cnt src_inc_gpc0(abuse-heavy-but-whitelisted) gt 0
  acl abuse_cnt src_get_gpc0(abuse) gt 0
  acl abuse_heavy_cnt src_get_gpc0(abuse-heavy) gt 0

  capture request header Host len 60
  capture request header User-Agent len 150
  capture request header Authorization len 60
  
  http-error status 429 content-type "text/html; charset=utf-8" lf-file /etc/haproxy/errors/429.html
  http-error status 502 errorfile /etc/haproxy/errors/502.http
  http-error status 503 errorfile /etc/haproxy/errors/503.http
  http-error status 504 errorfile /etc/haproxy/errors/504.http
  
  http-request del-header X-Forwarded-For
  http-request set-var(txn.body_size) req.body_size
  http-request set-header X-Forwarded-Host %[req.hdr(Host)]
  
  http-request track-sc0 src table http-request-rate-long
  http-request track-sc1 src table http-request-rate-short

  http-request capture url_param(auth_key) len 60 if { req.hdr_cnt(Authorization) eq 0 }

  http-request set-log-level notice

  http-request set-var(txn.ratelimitreason) str("00")
  http-request set-var(txn.special_treatment) int(0)
  http-request set-var(txn.special_treatment) int(1) if greylisted
  http-request set-var(txn.special_treatment) int(2) if whitelisted
  http-request set-var(txn.priority_class) int(0)

  http-request deny status 414 content-type "application/json" lf-file /etc/haproxy/errors/414.json if is_json uri_too_long
  http-request deny status 414 content-type "text/html; charset=utf-8" lf-file /etc/haproxy/errors/414.html if uri_too_long
  http-request deny status 420 content-type "text/html; charset=utf-8" lf-file /etc/haproxy/errors/429.html if is_blockedagent

  http-request set-var(txn.ratelimitreason) str("01")
  http-request silent-drop if abuse_heavy_cnt !whitelisted

  ttp-request set-var(txn.ratelimitreason) str("02")
  http-request silent-drop if heavy_abuser !whitelisted inc_abuse_heavy_cnt
  http-request silent-drop if greylisted !whitelisted heavy_abuser_if_greylisted inc_abuse_heavy_cnt

  http-request set-var(txn.ratelimitreason) str("03")
  http-request deny status 429 content-type "application/json" lf-file /etc/haproxy/errors/429.json if is_json abuse_cnt
  http-request deny status 429 content-type "text/html; charset=utf-8" lf-file /etc/haproxy/errors/429.html if abuse_cnt

  http-request set-var(txn.ratelimitreason) str("04")
  http-request deny status 429 content-type "application/json" lf-file /etc/haproxy/errors/429.json if is_json abuser !whitelisted inc_abuse_cnt
  http-request deny status 429 content-type "text/html; charset=utf-8" lf-file /etc/haproxy/errors/429.html if abuser !whitelisted inc_abuse_cnt
  http-request deny status 429 content-type "application/json" lf-file /etc/haproxy/errors/429.json if is_json greylisted !whitelisted abuser_if_greylisted inc_abuse_cnt
  http-request deny status 429 content-type "text/html; charset=utf-8" lf-file /etc/haproxy/errors/429.html if greylisted !whitelisted abuser_if_greylisted inc_abuse_cnt

  http-request set-var(txn.ratelimitreason) str("90") if abuser whitelisted inc_abuse_but_whitelisted_cnt
  http-request set-var(txn.ratelimitreason) str("91") if heavy_abuser whitelisted inc_abuse_heavy_but_whitelisted_cnt
  http-request set-log-level info

  http-request set-priority-class int(2) if low_prio_queue
  http-request set-priority-class int(2) if greylisted low_prio_queue_if_greylisted
  http-request set-priority-class int(1) if !low_prio_queue !greylisted || !low_prio_queue_if_greylisted
  http-request set-var(txn.priority_class) prio_class

  http-request redirect scheme https code 301 unless { ssl_fc }
  http-request set-header X-Forwarded-Proto https if { ssl_fc }
  http-request set-header X-Forwarded-Proto http if !{ ssl_fc }
  http-response set-header Strict-Transport-Security "max-age=63072000; includeSubDomains; preload"
  http-response set-log-level notice if { status ge 400 }

  log-format "%ci:%cp via %fi:%fp: %b/%s Status: %ST Cookies: %CC/%CS Termination: %ts Body size: %[var(txn.body_size)] Time: %Ta Request receive time: %TR Queue time: %Tw Upstream connect time: %Tc Response time: %Tr Retries: %rc Headers: %hr/%hs Proto: %HV Method: %HM URL: https://%[capture.req.hdr(0),json(ascii)]%HPO SSL cipher: %sslc Unique-ID: %[unique-id]"

  maxconn 3000000

  unique-id-format %{+X}o\ 01_%ci:%cp_%fi:%fp_%ST_%Ts_%[var(txn.special_treatment)]_%[var(txn.priority_class)]_%[var(txn.ratelimitreason)]
  unique-id-header X-Unique-ID

  # use_backend omitted (34x)...


frontend stats
  bind :9000 
  mode http
  http-request use-service prometheus-exporter if { path /metrics }
  http-request set-log-level silent if TRUE
  option dontlog-normal
  stats enable
  stats uri /stats
  stats refresh 10s

backend abuse
  stick-table type ip size 200K peers my_peers expire 5s store gpc0

backend abuse-but-whitelisted
  stick-table type ip size 200K peers my_peers expire 5s store gpc0

backend abuse-heavy
  stick-table type ip size 200K peers my_peers expire 300s store gpc0

backend abuse-heavy-but-whitelisted
  stick-table type ip size 200K peers my_peers expire 300s store gpc0

# backend omitted (34x)...

peers my_peers
  peer lb1.tld 10.0.0.101:9999
  peer lb2.tld 10.0.0.102:9999
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.6.6-1ppa1~focal 2022/09/22 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.6.html
Running on: Linux 5.4.0-128-generic #144-Ubuntu SMP Tue Sep 20 11:00:04 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-gKU17o/haproxy-2.6.6=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=48).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.4.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

Edit: I have just seen that I actually missed a few ACLs when swapping `src_` fetches with `sc_`:
```
  acl inc_abuse_cnt src_inc_gpc0(abuse) gt 0
  acl inc_abuse_heavy_cnt src_inc_gpc0(abuse-heavy) gt 0
  acl inc_abuse_but_whitelisted_cnt src_inc_gpc0(abuse-but-whitelisted) gt 0
  acl inc_abuse_heavy_but_whitelisted_cnt src_inc_gpc0(abuse-heavy-but-whitelisted) gt 0
  acl abuse_cnt src_get_gpc0(abuse) gt 0
  acl abuse_heavy_cnt src_get_gpc0(abuse-heavy) gt 0
```

Only the bottom two get evaluated each request. The others max 1reqs/s. This alone can not heave caused the heavy CPU usage, can it?

**Comment by capflam** [comment]
It seems you really hit some limitations about stick-tables. You may try the 2.7-dev8. @wtarreau recently improved this part. He observed a significant speed up on high load.

**Comment by phihos** [comment]
@capflam Sorry for the late answer. 

> You may try the 2.7-dev8. @wtarreau recently improved this part. He observed a significant speed up on high load.

So there have been improvements on the stick-tables apart from [this fix](https://github.com/haproxy/haproxy/commit/8bd146d8af78371f97b66e50cac718666eb93388) that throttled the stick-table peer sync?
If so this is good news. But I think I will wait until a stable release to try that out. Right now this is fixed by horizontal  scaling so that an instance only gets ~10k req/s.

**Comment by wtarreau** [comment]
Yes, the most important difference is the replacement of the spinlocks (exclusive locks) with an rwlock (shared lock) so that threads can work in parallel and do not cause contention anymore. As a rule of thumb I'd say that if you had 4 threads you would probably not notice a difference but with 43 you should :-) In my tests with 48 threads, the performance progressively increased from 183k to 2.07M, hence 11x! All this was merged into 2.7-dev8 (not quoting the series here, there are 12 patches or so).

**Comment by capflam** [comment]
FYI, the 2.7.0 was released

**Comment by capflam** [comment]
Any news about this issue ? @phihos, have you tested the 2.7 ?

**Comment by phihos** [comment]
Hi, with 2.7 HaProxy seems to be able handle the 20k rps just fine. I even did tests with about 40k rps and it worked as well. From my point of view you can close the issue.

**Comment by capflam** [comment]
Cool, thanks for the feedback. I'm closing

**Comment by stackbaek** [comment]
Curious - was this check backported to 2.2, or are only 2.7 and onward versions enjoy this?

**Comment by wtarreau** [comment]
The table's locking is extremely sensitive and relies on incremental architectural changes that were brought along versions. None of this was backported nor is backportable. Latest branch 3.0 is even way better than its predecessors regarding this, so if you think you're facing situations with high CPU usage in some stick-table functions, you might really want to consider giving 3.0 a try.

**Comment by stackbaek** [comment]
gotcha - thanks for the clarification!
